### DISCUSSION TITLE
CompletionStage VS CompletableFuture (vertx graphql - non blocking framework)

### DISCUSSION THREAD

**Discussion by knotenpunkt** [graphql-java/graphql-java#2289] [Ideas]
hey i saw in your code that you use CompletableFutures instead of the Completions-Stage

i'm not really sure if that is correct what i think about the following:

**The CompletableFuture is a blocking CompletionStage implementation, right?**

**But if you want to use it in non-blocking applications like vertx: https://vertx.io/docs/vertx-web-graphql/java/ or in some other framework, then they need to use a non-blocking CompletionStage implementation.**

So i took a look into the graphQL-implementation code and i found for example the following:

in ExecutionStrategy
```java
  try {
            Object fetchedValueRaw = dataFetcher.get(environment);
            fetchedValue = Async.toCompletableFuture(fetchedValueRaw);
        } catch (Exception e) {
            if (logNotSafe.isDebugEnabled()) {
                logNotSafe.debug(String.format("'%s', field '%s' fetch threw exception", executionId, executionStepInfo.get().getPath()), e);
            }

            fetchedValue = new CompletableFuture<>();
            fetchedValue.completeExceptionally(e);
```

and in Async

```java
    public static <T> CompletableFuture<T> toCompletableFuture(T t) {
        if (t instanceof CompletionStage) {
            //noinspection unchecked
            return ((CompletionStage<T>) t).toCompletableFuture();
        } else {
            return CompletableFuture.completedFuture(t);
        }
    }
```

so you create there instances of the blocking CompletableFuture, which will then not good work with the non-blocking-implementations.

**Wouldn't be that a good idea to refactor it to only CompletionStage? and for all cases in your graphql-lib where you need an instance of ANY implementation of CompletationStage (now its hard-coded to CompletableFuture) wouldn't be that there not a good idea having a interchangable factory, which can be replaced by frameworks like vertx? and in default cases it returns the blocking CompletableFuture?**

Anyway maybe there is also a solution not wrapping all fields into a instance of CompletionStage like fields that are directly available (     `return CompletableFuture.completedFuture(t);`) to save ressources and get a better performance?

**Comment by oliemansm** [comment]
I'm running into the same problem at the moment. While working on implementing support for timeout I ran into the problem that I couldn't cancel the `CompletableFuture` returned by `GraphQL.executeAsync`. This means I can cancel the HTTP request, but the GraphQL execution will still continue to run in the background which is undesirable. While debugging this issue further I noticed the same thing as explained by the reporter meaning that blocking operations of the `CompletableFuture` are currently used, making the entire thing blocking after all. I can't imagine that being the intent.

Updating the code to work with the `CompletionStage` instead of `CompletableFuture` might help in ensuring non-blocking functionality. That would change the public API too however, so I can imagine you might not want to move to it right away (although it might be cleaner). Ideally at least the existing logic is updated to become non-blocking while leaving the public API intact. 

See also: https://github.com/graphql-java-kickstart/graphql-spring-boot/issues/469#issuecomment-840399906.

**Reply by andimarek** [reply]
I think there is a misunderstanding: none of the code in GraphQL Java is blocking. The code you posted is not blocking.

CompletableFuture is a CompletionStage and every CompletionStage can be converted into a CompletableFuture via toCompletableFuture. They are in practical terms absolutely interchangeable.

Further: GraphQL Java never creates threads. This means problems like canceling a request thread is outside of the scope of this lib and you need to look into the other Libs you use on Top of GraphQL Java like Kickstarter. See https://www.graphql-java.com/blog/threads/ for more details around GraphQL Java and threads.

**Reply by oliemansm** [reply]
@andimarek 
Thanks for your quick response and sharing the link! It contained the critical information I was missing to make it work correctly:
> Specifically we piggyback on the `CompletableFuture` returned by the DataFetcher

The following just as a reference to anybody finding this discussion with the same question.

My original snippet to test timeouts was like this which became blocking because the `DataFetcher` was just returning a `String` directly:
```java
DataFetcher<String> test =
    env -> {
      Thread.sleep(3000);
      return "response";
    };
GraphQLSchema schema =
    GraphQLSchema.newSchema()
        .query(
            GraphQLObjectType.newObject()
                .name("query")
                .field(field -> field.name("test").type(Scalars.GraphQLString))
                .build())
        .codeRegistry(
            GraphQLCodeRegistry.newCodeRegistry()
                .dataFetcher(FieldCoordinates.coordinates("query", "test"), test)
                .build())
        .build();
GraphQL graphql = GraphQL.newGraphQL(schema).build();
log.info("#1 - Run query");
CompletableFuture<Void> future =
    graphql.executeAsync(ExecutionInput.newExecutionInput().query("{ test }").build())
    .thenAccept(it -> log.info("#3 - Result: {}", it));
log.info("#2 - Future: {}", future);
```
The logs showed the process was blocking because the order is off:
```
2021-05-13 12:29:53.928  INFO 300768 --- [           main] #1 - Run query
2021-05-13 12:29:57.018  INFO 300768 --- [           main] #3 - Result: ExecutionResultImpl{errors=[], data={test=response}, dataPresent=true, extensions=null}
2021-05-13 12:29:57.019  INFO 300768 --- [           main] #2 - Future: java.util.concurrent.CompletableFuture@7a6ea47d[Completed normally]
```
Once I changed the `DataFetcher` to a `CompletableFuture` like this:
```java
    DataFetcher<CompletableFuture<String>> test =
        env ->
            CompletableFuture.supplyAsync(
                () -> {
                  try {
                    Thread.sleep(3000);
                  } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                  }
                  return "response";
                });
```
It indeed became non-blocking:
```
2021-05-13 12:44:06.629  INFO 302486 --- [           main] #1 - Run query
2021-05-13 12:44:06.691  INFO 302486 --- [           main] #2 - Future: java.util.concurrent.CompletableFuture@6cbbb9c4[Not completed]
2021-05-13 12:44:09.693  INFO 302486 --- [onPool-worker-5] #3 - Result: response
```

**Comment by andimarek** [comment]
See my comment above: https://github.com/graphql-java/graphql-java/discussions/2289#discussioncomment-733054
